Add support for Generics (requires Go 1.18+)

hawkinsw opened the pull request:
This patch represents a first attempt at adding support for generics.

Thank you @dave for the amazing work on this tool! The code is really fun to work with! I am sure that there are many other things that will need to be done to get this going, but I made pretty good progress today and wanted to let you take a look!

Thanks for letting me help!
Will

Towards #63

dave commented:
Wow this looks awesome. Take a look at this file: https://github.com/dave/dst/blob/master/gendst/data/positions.go

... at some point you'll need to add the type parameters to it and work out where valid decorations (e.g. inline comments) can be placed... I did that bit my putting inline comments absolutely everywhere, and running `gofmt` ... what you'll see is some positions won't allow comments, and they'll be moved to other locations... That way you can get a comprehensive list of all the decoration locations, and give them sensible names. That file is then used by a test ([here I think](https://github.com/dave/dst/blob/4f484ea8fc7f5a6532bd5b227389d369eff6a332/decorator/decorator_pos_test.go#L20)) that checks they all work...

dave commented:
That file `positions.go` also gets sliced up automagically to make some of the documentation I think... (see [here](https://github.com/dave/dst/blob/master/gendst/README.md))

dave commented:
> The code is really fun to work with

Thanks! Without all the code generation this project would have been a lot more work!

hawkinsw commented:
> Wow this looks awesome. Take a look at this file: https://github.com/dave/dst/blob/master/gendst/data/positions.go
> 
> ... at some point you'll need to add examples of all variations of type parameters nodes to it and work out where valid decorations (e.g. inline comments) can be placed... I did that bit my putting inline comments everywhere (e.g. between absolutely every token), and running `gofmt` ... what you'll see is some positions won't allow comments, and they'll be moved to other locations... That way you can get a comprehensive list of all the valid decoration locations, and give them sensible names. That file is then used by a test ([here I think](https://github.com/dave/dst/blob/4f484ea8fc7f5a6532bd5b227389d369eff6a332/decorator/decorator_pos_test.go#L20)) that checks they all work...

I *sincerely* appreciate your enthusiastic and quick reply! That honestly makes it all worth it! I will take a peak at positions.go and work to make the additions. 

Do you want those to be cumulative or just complete? For instance, you can define a function with and without type parameters. So, would you want two examples (one with type parameters and one without) or just a single example that uses type parameters? 

I hope that question makes sense!

Thanks again -- I am starting on the positions.go file as we speak!

hawkinsw commented:
Just FYI: I worked through some issues that arose as a result of changes in `positions.go` and ran into a few corner cases that the code does not properly handle. I hoped to have another PR ready for you to review tonight, but I am sorry to say that it will take another few hours of work (at least!). I am really sorry!

Will

hawkinsw commented:
Sunday update, @dave : It looks like what we really need is a new `dst` type that does not necessarily correspond to an `ast` type if we are going to make this work really well. The `ast` treats type parameters as just a `FieldList` but that does not give the dst user enough flexibility for annotations, IMO. In order to *really* make the `dst` user happy, I am going to make a special `TypeParams` `dst` node. That will give the `dst` user the option to do lots more annotations (I think). It will require some additional surgery but, in the long run, I think that it will be much better. 

Let me know what you think about this approach?

Thanks again for everything!
Will

dave commented:
Are you sure this is needed? `dst.FuncType.Params` and `dst.FuncType.Results` are just `FieldList`... What is it about `TypeParams` that makes it need a spacial type?

hawkinsw commented:
Yes, I think that you are 100% right! The more that I thought about it the
more I realized that I don't think we should go that route. So, I am
backing out and doing some changes to positions.go now that I hope will get
us right!

Sorry for the over engineering!

Will


On Mon, May 9, 2022 at 12:31 PM Dave Brophy ***@***.***>
wrote:

> Remember the goal of the dst syntax tree is to be as close as possible to
> the ast syntax tree but preserve comments and line spacing... Not invent
> a better syntax tree...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dave/dst/pull/66#issuecomment-1121321753>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACCP2CQG5AZLMUINK2KBYLTVJE4UXANCNFSM5VJ62XQQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

dave commented:
🙏💯

hawkinsw commented:
> pray 100

@dave I think I have addressed your feedback for some additional work on `positions.go`. Let me know what I missed -- I am *sure* that there is something!

Thanks again for being so welcoming and inviting! 

Will

dave commented on gendst/data/positions.go:
```diff
@@ -85,32 +85,32 @@ var Q = map[string]string{
 
 // ArrayType
-type R /*Start*/ [ /*Lbrack*/ 1] /*Len*/ int /*End*/
+type R[P any] /*Start*/ [ /*Lbrack*/ 1] /*Len*/ P /*End*/
```
Are all the type parameters needed on these examples? I think it's better to leave them as simple as possible...

hawkinsw replied:
Absolutely understood! I was just trying to make them complete -- I didn't know which one you preferred!

dave commented on gendst/data/positions.go:
```diff
@@ -280,7 +280,12 @@ A /*Label*/ : /*Colon*/
 	// TypeSpec(1)
 	type (
-		/*Start*/ T2 = /*Name*/ T1 /*End*/
+		/*Start*/ T2 = /*AliasName*/ T1 /*End*/
```
Was it intentional to rename `Name` to `AliasName`? This would be a breaking change...

hawkinsw replied:
Unfortunately, it *was* intentional. I didn't necessarily realize that it would be a breaking change. Here's the problem:

```
type (
/*start*/ T[P any] /*Name*/ []P
)
```
is the only way decorations will work in that case because the `Name` decoration has to come after the set of type parameters because the `Name` decoration goes after the `=` in an AliasDecl. 

In other words, we cannot do this:
```type (
/*Start*/ T /*Name*/ [P any] /*TypeParams*/ []P
)
```

I am not 100% sure whether that is something you want to support, but that is where the `AliasName` came from -- a way to differentiate the decorator. 

I am sorry if I am not explaining clearly. Please let me know if I confused the situation! I am happy to resolve it however you deem appropriate -- you are the expert!

dave replied:
What do you think about this?

```go
	// TypeSpec(0)
	type (
		/*Start*/ T1 /*Name*/ []int /*End*/
	)

	// TypeSpec(1)
	type (
		/*Start*/ T2 = /*Name*/ T1 /*End*/
	)

	// TypeSpec(2)
	type (
		/*Start*/ T3 /*TypeParams*/ [P any] = /*Name*/ []P /*End*/
	)
```

dave replied:
Would that work? I imagine putting a comment between the name and the type parameters will be pretty rare...

hawkinsw replied:
Interestingly enough, `TypeSpec(2)` is invalid because it is not possible to have a generic alias. So, that might not be a problem at all! 

This, however, does not work:

```
type (
/*Start*/ T3 /*TypeParams*/ [P any] /*Name*/ []P /*End*/
)
```

dave replied:
OK cool that works. I haven't actually played with Go generics on a project yet 😂

dave replied:
Oh sorry I read "This, however, does work:".... how doesn't it work?

hawkinsw replied:
No problem! It doesn't work because it wants /*End*/ after T3 because it considers that the _end_ of an `Ident`. 

Sorry for the delay responding!

dave replied:
Hmmm I think if that's the problem, then we have something wrong in the setup... there will be an ident in there which will have start and end annotation, but the additional annotation should work just fine... take a look how similar nodes work... 🤔

dave replied:
e.g. take a look at CallExpr.Fun - that's the same scenario, right?

hawkinsw replied:
Your suggestion is perfect!!

dave commented on gendst/data/positions.go:
```diff
@@ -280,7 +280,12 @@ A /*Label*/ : /*Colon*/
+
+	// TypeSpec(2)
+	type (
+		X /*Name*/ [P any] /*TypeParams*/ []P
```
I think we're missing the `Start` and `End` decorations...

hawkinsw replied:
Good :eye: !

dave commented on go.mod:
```diff
@@ -1,16 +1,17 @@
 module github.com/dave/dst
```
There seems to be a lot more changes to `go.mod` than I'd expect... Can we revert them without breaking things?

hawkinsw replied:
You are 100% correct. I was fumbling around early on in my work and didn't realize how everything worked together. I think that I have it cleaned up now so that the change to these files are not required. Of course, I cannot test outside of what I have installed, so please let me know if I did something wrong!

dave commented on dst.go:
```diff
@@ -258,10 +258,11 @@ type (
-		Results *FieldList // (outgoing) results; or nil
-		Decs    FuncTypeDecorations
+		Func       bool
+		TypeParams *FieldList
```
Let's copy the comment from [ast.go](https://github.com/golang/go/blob/master/src/go/ast/ast.go#L450)

dave commented on gendst/data/data.go:
```diff
@@ -1961,7 +1982,8 @@ var Info = map[string][]Part{
 		Decoration{
-			Name: "Name",
+			Name: "AliasName",
+			Use:  Expr(func(n *jen.Statement) *jen.Statement { return n.Dot("Assign").Dot("IsValid").Call() }),
```
Was adding `Use` intentional? I'm hazy on how this works, but it doesn't feel like this was needed. As before renaming `Name` will break backwards compatibility, so I assume it was a mistake?

hawkinsw replied:
See above.

dave commented on gendst/data/data.go:
```diff
@@ -2135,6 +2157,20 @@ var Info = map[string][]Part{
+			Name: "TypeParams",
+			Use:  Expr(func(n *jen.Statement) *jen.Statement { return n.Dot("Type").Dot("TypeParams").Op("!=").Nil() }),
+		},
+		SpecialDecoration{
```
Was the use of `SpecialDecoration` intentional?

hawkinsw replied:
It was. Do I misunderstand the use of the `SpecialDecoration` flag here? Like `Recv`, there is no reason to offer that decoration of `TypeParams` if there are no `TypeParams` defined. Again, please educate me -- you are the expert!

dave replied:
Haha I'm a little hazy on how it works! I'd have to read through the code...

dave replied:
If it was intentional you're probably correct.

dave reviewed (commented):
Looks great! I've added some comments...

hawkinsw commented:
Hello @dave -- thanks (as usual) for your eager engagement! I just pushed some new updates. They do not address the issue of `AliasName` but I *believe* that they address everything else! Please let me know what I missed!

dave commented on gendst/data/positions.go:
```diff
@@ -97,7 +97,7 @@ type T /*Start*/ func /*Func*/ (a int) /*Params*/ (b int) /*End*/
 
```
In here we need something to show FuncType.TypeParams, like this:

```go
// FuncType(0)
type T1 /*Start*/ func /*Func*/ (a int) /*Params*/ (b int) /*End*/

// FuncType(1)
func F1 /*Start*/ [P any] /*TypeParams*/ (p P) /*Params*/ P/*End*/ { return p }
```

dave replied:
No reason to be sorry... better to think these things out and take time than rush.

hawkinsw replied:
Example added!

hawkinsw commented:
@dave -- Try that!

dave commented on gendst/data/data.go:
```diff
@@ -835,9 +835,10 @@ var Info = map[string][]Part{
-			Params  *FieldList // (incoming) parameters; non-nil
-			Results *FieldList // (outgoing) results; or nil
+			Func    token.Pos     // position of "func" keyword (token.NoPos if there is no "func")
+		  TypeParams *FieldList // type parameters; or nil
```
Tabs vs spaces issue here.

dave commented on gendst/data/data.go:
```diff
@@ -1934,11 +1944,12 @@ var Info = map[string][]Part{
-			Comment *CommentGroup // line comments; or nil
+			Doc        *CommentGroup // associated documentation; or nil
+			Name       *Ident        // type name
+		  TypeParams *FieldList    // type parameters; or nil
```
Again here

hawkinsw replied:
Well, I suck! Sorry!

dave commented on gendst/data/positions.go:
```diff
@@ -92,12 +92,12 @@ type S /*Start*/ struct /*Struct*/ {
 // InterfaceType
 type U /*Start*/ interface /*Interface*/ {
-	A()
+	A(a int)
```
I don't think this is needed

hawkinsw replied:
Correct! Sorry again!

hawkinsw commented:
Hope this does the trick! @dave

stapelberg commented:
Thanks for working on this. I’m also interested in using dst with code that uses generics, and I can confirm that this PR seems to work (at least with anything I have tried so far) :)

stapelberg commented:
Spoke too soon, sorry!

The following program makes the `dst` package panic (with this PR applied):
```go
package main

type FanOut[Source any, Payload any] struct {
	src Source
}

type Receiver[Source any, Payload any] struct {
	src FanOut[Source, Payload]
}

func main() {
	println("Hello, World!")
}
```

I think the issue is in line `src FanOut[Source, Payload]`, which uses multiple type parameters. If I use `src FanOut[Source]`, there is no panic.

Here’s a full reproducer (save as dstpanic_test.go): https://gist.github.com/stapelberg/4ad06a53ef24c833c9b6eccf97833926

Running it gives me this output:
```
go test -v 
=== RUN   TestGenerics
=== RUN   TestGenerics/AST
=== RUN   TestGenerics/DST
--- FAIL: TestGenerics (0.00s)
    --- PASS: TestGenerics/AST (0.00s)
    --- FAIL: TestGenerics/DST (0.00s)
panic: interface conversion: interface is nil, not dst.Expr [recovered]
	panic: interface conversion: interface is nil, not dst.Expr

goroutine 8 [running]:
testing.tRunner.func1.2({0x6e91a0, 0xc0001cced0})
	/home/michael/sdk/go1.18/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
	/home/michael/sdk/go1.18/src/testing/testing.go:1392 +0x39f
panic({0x6e91a0, 0xc0001cced0})
	/home/michael/sdk/go1.18/src/runtime/panic.go:838 +0x207
github.com/dave/dst/decorator.(*fileDecorator).decorateNode(0xc0001cbdc0, {0x798838, 0xc00009cde0}, {0x72ed85, 0x9}, {0x72da12, 0x4}, {0x72dc7d, 0x5}, {0x798810, ...})
	/home/michael/go/src/github.com/hawkinsw/dst/decorator/decorator-node-generated.go:706 +0x1332e
github.com/dave/dst/decorator.(*fileDecorator).decorateNode(0xc0001cbdc0, {0x798c70, 0xc00000c138}, {0x72f4c9, 0xa}, {0x72dfe5, 0x6}, {0x72ed85, 0x9}, {0x798838, ...})
	/home/michael/go/src/github.com/hawkinsw/dst/decorator/decorator-node-generated.go:744 +0x124e5
github.com/dave/dst/decorator.(*fileDecorator).decorateNode(0xc0001cbdc0, {0x798ce8, 0xc000010fc0}, {0x72e9c9, 0x8}, {0x72da66, 0x4}, {0x72d9de, 0x4}, {0x798c70, ...})
	/home/michael/go/src/github.com/hawkinsw/dst/decorator/decorator-node-generated.go:2003 +0x9a9d
github.com/dave/dst/decorator.(*fileDecorator).decorateNode(0xc0001cbdc0, {0x798928, 0xc000011100}, {0x72e3f4, 0x7}, {0x72dd0e, 0x5}, {0x72da56, 0x4}, {0x798ce8, ...})
	/home/michael/go/src/github.com/hawkinsw/dst/decorator/decorator-node-generated.go:2166 +0xa9c5
github.com/dave/dst/decorator.(*fileDecorator).decorateNode(0xc0001cbdc0, {0x798860, 0xc0000ca280}, {0x72d9e6, 0x4}, {0x72dc5a, 0x5}, {0x72d9d2, 0x4}, {0x798928, ...})
	/home/michael/go/src/github.com/hawkinsw/dst/decorator/decorator-node-generated.go:1109 +0x11efd
github.com/dave/dst/decorator.(*fileDecorator).decorateNode(0xc0001cbdc0, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x798860, ...})
	/home/michael/go/src/github.com/hawkinsw/dst/decorator/decorator-node-generated.go:788 +0x11b25
github.com/dave/dst/decorator.(*Decorator).DecorateNode(0xc0000edea0, {0x798860?, 0xc0000ca280?})
	/home/michael/go/src/github.com/hawkinsw/dst/decorator/decorator.go:134 +0x185
github.com/dave/dst/decorator.(*Decorator).DecorateFile(...)
	/home/michael/go/src/github.com/hawkinsw/dst/decorator/decorator.go:109
dstpanic_test.TestGenerics.func2(0xc0000c7380)
	/home/michael/go/src/dstpanic/dstpanic_test.go:48 +0x18b
testing.tRunner(0xc0000c7380, 0xc00009d380)
	/home/michael/sdk/go1.18/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/home/michael/sdk/go1.18/src/testing/testing.go:1486 +0x35f
exit status 2
FAIL	dstpanic	0.004s
```

dave commented:
Awesome, thanks @stapelberg! 

@hawkinsw I haven't had time to take a look at this, but I'm expecting to make some time this weekend to have a play.

hawkinsw commented:
@stapelberg and @dave -- I have isolated the problem! The fix should be relatively straightforward (famous last words!) I should have time to work on it tomorrow after ${DAYJOB}. 

Sorry for the delay!

Will

stapelberg commented:
Great, thanks for taking a stab at this :)

hawkinsw commented:
@dave @stapelberg Check out the newest rev of this patch. I *believe* that I solved the problem!

Sorry again for the delay!

dave commented:
I'm just starting to look through this... The first thing I'm doing is going through the [commit history for ast.go](https://github.com/golang/go/commits/master/src/go/ast/ast.go) to make sure we haven't missed any updates... looks like there's a few comment updates we haven't copied over to `dst.go` - the first couple I found:

The updated comment on FieldList (dst.go line 73) from: https://github.com/golang/go/commit/83bfdb6561c06bd890446af01b0f0c735360a8df

The updated comments on Field, Field.Names and Field.Type (dst.go line 65) from:
https://github.com/golang/go/commit/925ea2dfc4911188d47cca5c5534916e39dc1463

... I guess there's more we're missing...

hawkinsw commented:
> I'm just starting to look through this... The first thing I'm doing is going through the [commit history for ast.go](https://github.com/golang/go/commits/master/src/go/ast/ast.go) to make sure we haven't missed any updates... looks like there's a few comment updates we haven't copied over to `dst.go` - the first couple I found:
> 
> The updated comment on `FieldList` (dst.go line 73) from: [golang/go@83bfdb6](https://github.com/golang/go/commit/83bfdb6561c06bd890446af01b0f0c735360a8df)
> 
> The updated comments on `Field`, `Field.Names` and `Field.Type` (dst.go line 65) from: [golang/go@925ea2d](https://github.com/golang/go/commit/925ea2dfc4911188d47cca5c5534916e39dc1463)
> 
> ... I guess there's more we're missing...

I went back to the commit(s) in `src/go/ast/ast.go` from (late) March 2020 (which matched the last time that `gendst/data/data.go` was updated and looked at all the comment changes. I believe that we are now in sync! Let me know what you think!

(See how that rhymed?)

stapelberg commented:
Can confirm that with the current pull request contents, the issue I ran into in https://github.com/dave/dst/pull/66#issuecomment-1129901348 is fixed!

Thanks :)

hawkinsw commented:
> Can confirm that with the current pull request contents, the issue I ran into in [#66 (comment)](https://github.com/dave/dst/pull/66#issuecomment-1129901348) is fixed!
> 
> Thanks :)

Thank you for the confirmation, @stapelberg !!

stapelberg commented:
@dave friendly ping? Is this PR good to go now? :)

dave commented:
Hey sorry I've been hiking in the mountains for the last few days... should have some time to take another look at this over the next few days...

hawkinsw commented:
> Hey sorry I've been hiking in the mountains for the last few days... should have some time to take another look at this over the next few days...

Hope that you are enjoying it!!

dave commented:
Hey so I had a good play with this, added some more tests, found some more bugs, fixed them, merged and released it as v0.27.0. Thanks so much for this @hawkinsw!
